### PR TITLE
Added support for public clients

### LINKED ISSUE #1714: Client secret is not required for OIDC with PKCE enabled
First off, thanks for the great project!

## Expected Behavior

I have configured oauth2-proxy with Okta and configured the Okta app to be a public client using PKCE and in this case there is no client secret provided by Okta. It would be great to be able to start oauth2-proxy without a client secret because it works perfectly with PKCE if you are able to work around not supplying the client secret.

## Current Behavior

The startup validation checks that client secret is provided unless the provider is login.gov:

https://github.com/oauth2-proxy/oauth2-proxy/blob/f6aa7600ea867ec069f86990501c0629aa4b5064/pkg/validation/providers.go#L51-L61

However, I've found that it's possible to work around this by supplying an empty client secret file via the client-secret-file argument.

## Possible Solution

Perhaps a special value of 'none' would skip validation - or maybe tie it to the code challenge type parameter being set (although presumably some OIDC providers do provide a client secret when using pkce?)

## Steps to Reproduce (for bugs)

Start oauth2 proxy without a client secret supplied, startup will error.

1.  Set client secret to an empty string
2.  Set other params to use Okta OIDC with PKCE enabled

```
      image:
        tag: v7.3.0
      config:
        clientID: blahblahblah
        clientSecret: ""
        cookieSecret: blahblahblahblahblah
      extraArgs:
        provider: oidc
        redirect-url: /oauth2/callback
        oidc-issuer-url: https://somedomain.oktapreview.com/oauth2/default
        code-challenge-method: S256
        email-domain: "*"
        skip-provider-button: true
        set-authorization-header: true
        set-xauthrequest: true
        scope: openid profile email groups
```

4.  Start oauth2-proxy

## Context

Trying to use oauth2-proxy with Okta PKCE as per (this)[https://developer.okta.com/blog/2019/08/22/okta-authjs-pkce] article.

## Your Environment

- Version used:

7.3.0

### PR DESCRIPTION
**Pull request by jrote1**
## Description

Adds support for public clients

## Motivation and Context

Fixes #1714

## How Has This Been Tested?

I have tested this locally for a public client app that we use and works as expected as the only change required is to not to send a secret as part of the token exchange after auth

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by jrote1** [comment]
Would you like me to update any of the docs for this change?

**Comment by nickytd** [comment]
May I ask for comments on this PR? It is still valid.

**Comment by jrote1** [comment]
@nickytd Hi, I am still using this code change in or environment and works perfectly. I would like to see this pulled in and am happy to help where needed to get this pulled in

**Comment by nickytd** [comment]
Yes me too. I also run this in production. Hope project committers can provide feedback and we will respect it in the proposal.

File: pkg/validation/providers.go [hunk 1]
```diff
@@ -48,7 +48,7 @@ func validateProvider(provider options.Provider, providerIDs map[string]struct{}
 
 	// login.gov uses a signed JWT to authenticate, not a client-secret
 	if provider.Type != "login.gov" {
-		if provider.ClientSecret == "" && provider.ClientSecretFile == "" {
+		if provider.ClientSecret == "" && provider.ClientSecretFile == "" && provider.Type != "oidc" {
```
Comment (miguelborges99) [inline pkg/validation/providers.go:51]: OIDC may need client secret, if it uses confidential clients. I think it is optional if PKCE is used (option --code-challenge-method)

**Comment by Tawmu** [comment]
This is still a problem

**Comment by koen92** [comment]
A workaround for this issue is to use `/dev/null` as client secret file, for instance by passing this parameter to oauth2-proxy:

```
oauth2-proxy --client-secret-file /dev/null ....
```

Then it'll ignore the client secret and act as a public client.

**Pull request closed without merging**